### PR TITLE
[: install test as [ as well

### LINKED ISSUE #2420: make with MULTICALL is missing symlink from test to [
When trying to build ``i3`` we encountered the following error from meson:

```
meson.build:128:60: ERROR: Program or command '[' not found or not executable
```

which made me realize that the symlink from ``test`` to ``[`` is not created when calling the makefile with MULTICALL.

### PR DESCRIPTION
**Pull request by miDeb**
Closes #2420.

This installs `test` as `[` as well. When `test` is called with a binary name that ends with `[`, it changes its behavior a bit.
I would rather add `[` as its own utility instead of changing the behavior of `test` based on its binary name,
but I ran into problems trying to do that because `[` is not a valid crate/feature name. Working around it in build.rs and in GNUmakefile adds complexity and feels easy to get wrong... In any case we can't convince cargo to create a binary called `[`, which we can't work around if `uutils` is installed via `cargo install`.

There are plans for cargo to [support creating arbitrarily named binaries](https://github.com/rust-lang/cargo/issues/1706), which would mean we'd only have to work around the restriction for the feature name.

Maybe we can solve this problem in some other way?

### CODE REVIEW

**Comment by jhscheer** [comment]
> I would rather add `[` as its own utility instead of changing the behavior of `test` based on its binary name,
> but I ran into problems trying to do that because `[` is not a valid crate/feature name.

I think your approach as it is right now is a valid workaround.

The manpage speaks of another subtle difference:
> NOTE:  [  honors the --help and --version options, but test does not.  test treats each of those as it treats any other nonempty STRING.

**Comment by miDeb** [comment]
> NOTE: [ honors the --help and --version options, but test does not. test treats each of those as it treats any other nonempty STRING.

Thanks for pointing this out! Are we allowed to copy the help text from GNU, or do we have to come up with our own?

File: src/uu/test/src/test.rs [hunk 1]
```diff
@@ -10,12 +10,81 @@
...
 use std::path::Path;
 use uucore::executable;
 
+const USAGE: &str = "test EXPRESSION
+or:  test
+or:  [ EXPRESSION ]
+or:  [ ]
+or:  [ OPTION";
+
+const AFTER_HELP: &str = "
```
Comment (sylvestre) [inline src/uu/test/src/test.rs:25]: please add a comment explaining why this

File: src/uu/test/src/test.rs [hunk 2]
```diff
@@ -31,7 +100,18 @@ pub fn uumain(mut args: impl uucore::Args) -> i32 {
...
     // If invoked via name '[', matching ']' must be in the last arg
-    if binary_name == "[" {
+    if binary_name.ends_with('[') {
+        if args.len() == 1 && (args[0] == "--help" || args[0] == "--version") {
+            // Let clap pretty-print help and version
+            App::new(binary_name)
+                .version(crate_version!())
+                .usage(USAGE)
+                .after_help(AFTER_HELP)
+                .setting(AppSettings::NeedsLongHelp)
```
Comment (sylvestre) [inline src/uu/test/src/test.rs:110]: please add a comment to explain why these two settings :)

File: src/uu/test/src/test.rs [hunk 3]
```diff
@@ -10,12 +10,81 @@
...
+or:  [ ]
+or:  [ OPTION";
+
+const AFTER_HELP: &str = "
+Exit with the status determined by EXPRESSION.
+
+An omitted EXPRESSION defaults to false.  Otherwise,
+EXPRESSION is true or false and sets exit status.  It is one of:
+
+  ( EXPRESSION )               EXPRESSION is true
```
Comment (sylvestre) [inline src/uu/test/src/test.rs:31]: all this will replace the current help mgmt or add to it?

Comment (miDeb) [reply src/uu/test/src/test.rs:31]: test does not handle the help flag at all. Only when installed as [ it should handle --help and --version, which I'm adding here. This is basically the first help mgmt test got, there was none before.

Comment (sylvestre) [reply src/uu/test/src/test.rs:32]: sounds great, thanks :)

**Comment by SuperSandro2000** [comment]
I am on 2177b8dc37d9c8705b3c59efe7c5e10a69dce182 and the alias is not installed for me.

**Pull request merged**